src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -336,11 +339,40 @@ internal IEnumerable<IEdmNavigationProperty> GetSelectedNavigationProperties(IEd
             return selectedNavigationProperties.Distinct();
         }
 
+        /// <summary>
+        /// Gets all selected properties at the current node.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The selected properties at this node level.</returns>
+        internal IEnumerable<IEdmProperty> GetSelectedProperties(IEdmStructuredType structuredType)
+        {
+            if (this.selectionType == SelectionType.Empty)
+            {
+                return EmptyEdmProperties;
+            }
+
+            // We cannot determine the selected properties without the user model. This means we won't be computing the missing properties.
+            if (structuredType == null)
+            {
+                return EmptyEdmProperties;
+            }
+
+            if (this.selectionType == SelectionType.EntireSubtree || this.hasWildcard)
+            {
+                return structuredType.DeclaredProperties;
+            }
+
+            Debug.Assert(this.selectedProperties != null, "selectedProperties != null");
+
+            // Get declared properties selected, and filter out unrecognized properties.
+            return this.selectedProperties.Select(structuredType.FindProperty).OfType<IEdmProperty>();

```

Comment:
Why are we omitting undefined properties? If a dynamic property is explicitly requested in the select list, I would expect to return a null value for it if it was a null value or omitted in the payload.

Reply:
good point. Will do.

---
In reply to: [186531887](https://github.com/OData/odata.net/pull/1106#discussion_r186531887) [](ancestors = 186531887)